### PR TITLE
Improve api delete

### PR DESCRIPTION
**Pull request by AmaliMatharaarachchi**
Fixes https://github.com/wso2-enterprise/choreo/issues/7438
Support partial deletions

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 1]
```diff
@@ -3436,52 +3427,92 @@ public void deleteAPI(API api) throws APIManagementException {
                                 + " successfully removed from the database.";
                 log.debug(logMessage);
             }
+        } catch (APIManagementException e) {
+            log.error("Error while executing API delete operations on DB", e);
```
[the remark this answers is not in the document]
Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:3431]: Lets include it in all error logs

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 2]
```diff
@@ -3436,52 +3427,92 @@ public void deleteAPI(API api) throws APIManagementException {
...
-            if (recommendationEnvironment != null) {
-                RecommenderEventPublisher
-                        extractor = new RecommenderDetailsExtractor(api, tenantDomain, APIConstants.DELETE_API);
-                Thread recommendationThread = new Thread(extractor);
-                recommendationThread.start();
-            }
+        // if one of the above has failed throw an error
+        if (error) {
+            throw new APIManagementException("Error while deleting the API " + api.getUuid());
+        }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:3500]: Wondering whether we should still send the latter deletion events etc during an error in previous steps.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AmaliMatharaarachchi) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: Added API uuid to logs with https://github.com/wso2/carbon-apimgt/pull/10727/commits/0d0876a2ba34bcfa6977c45a1f13f3ff3bb7b5f5

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 3]
```diff
@@ -3436,52 +3427,98 @@ public void deleteAPI(API api) throws APIManagementException {
...
-                        }
+        if (apiId != -1) {
+            try {
+                cleanUpPendingAPIStateChangeTask(apiId);
+            } catch (WorkflowException | APIManagementException e) {
+                log.error("Error while executing API delete operation on cleanup workflow tasks for API "
+                        + api.getUuid(), e);
+                isError = true;
+            }
+        }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:3469]: Deprioritize and put it at the bottom of the list

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 4]
```diff
@@ -3436,52 +3427,98 @@ public void deleteAPI(API api) throws APIManagementException {
...
+            if (apiStoreSet != null && !apiStoreSet.isEmpty()) {
+                for (APIStore store : apiStoreSet) {
+                    wso2APIPublisher.deleteFromStore(api.getId(), APIUtil.getExternalAPIStore(store.getName(), tenantId));
+                }
+            }
+        } catch (APIManagementException e) {
+            log.error("Error while executing API delete operation on external API stores for API "
+                    + api.getUuid(), e);
+            isError = true;
+        }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:3451]: We can deprioritize and put it at the bottom of the list

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 5]
```diff
@@ -3436,52 +3427,97 @@ public void deleteAPI(API api) throws APIManagementException {
...
-            APIUtil.logAuditMessage(APIConstants.AuditLogConstants.API, apiLogObject.toString(),
-                    APIConstants.AuditLogConstants.DELETED, this.username);
+        JSONObject apiLogObject = new JSONObject();
+        apiLogObject.put(APIConstants.AuditLogConstants.NAME, api.getId().getApiName());
+        apiLogObject.put(APIConstants.AuditLogConstants.VERSION, api.getId().getVersion());
+        apiLogObject.put(APIConstants.AuditLogConstants.PROVIDER, api.getId().getProviderName());
 
-            cleanUpPendingAPIStateChangeTask(apiId);
+        APIUtil.logAuditMessage(APIConstants.AuditLogConstants.API, apiLogObject.toString(),
+                APIConstants.AuditLogConstants.DELETED, this.username);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:3441]: shall we move this to just after priority deletion tasks are completed?
we can do it just before
```
        if (apiId != -1) {
            try {
                cleanUpPendingAPIStateChangeTask(apiId);
```

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**